packages/core/src/model.types.ts

```diff
@@ -12,7 +12,18 @@ import {
 
 export type AnyFunction = (...args: any[]) => any;
 
-export type Cast<A1 extends any, A2 extends any> = A1 extends A2 ? A1 : A2;
+// https://github.com/microsoft/TypeScript/issues/23182#issuecomment-379091887
+export type IsNever<T> = [T] extends [never] ? true : false;
+
+/**
+ * Casts T to TCastType if T is assignable to TCastType.
+ * When `never` extends `TCastType`, `TCastType` will be chosen to prevent `never`.
+ */
+export type Cast<T extends any, TCastType extends any> = IsNever<T> extends true
```

Comment:
@Andarist @devanshj Can you please double check this?

My VS Code was showing errors (though mysteriously, CI was passing) because `Cast<never, EventObject>` resulted in `never`, when I would expect it to result in `EventObject`. This is because `never extends EventObject` (which still confuses me) but I added a check to make sure this doesn't happen.

Reply:
The short answer is I'd very strongly suggest to not change `Cast` because that's not the problem here and it's behaving perfectly fine. I'd suggest you to check for `never` at the place of issue and then swap it with `EventObject`.

But the fact that you need a check like this is weird can you tell me the original problem? Where exactly `Cast<never, EventObject>` being `never` was a problem? I'll clone the repo

I'm sure you know some of this but we use Cast in places like...
```ts
type Foo<T extends string> = {}
type Bar<T> = Foo<Cast<T, string>>
// Foo<T> doesn't match the constraint
```

But if you have...
```ts
type Foo<T extends string> = {}
type Bar<T extends never> = Foo<T>
// no problemo no point of using Cast and even if you use it Cast<never, string> will be never
```

So the fact you need your version of `Cast` itself is fishy

Reply:
For reference, the problem happens here: https://github.com/statelyai/xstate/pull/2266#issuecomment-885206683

Reply:
@devanshj The reason we need `Cast` here is because of this:

![CleanShot 2021-07-23 at 18 15 09](https://user-images.githubusercontent.com/1093738/126846504-b32fae41-f823-46e6-bea7-0f1a702a21da.png)

> Type 'TComputedEvent' does not satisfy the constraint 'EventObject'.ts(2344)

Even though _we_ know that `TComputedEvent` will always be an `EventObject`, TS doesn't seem to know this.

Reply:
Yep I understand the original requirement of cast, that's valid. But wanting `Cast<never, EventObject>` to be `EventObject` (which is not "casting" behavior as `never` already suffices) is invalid. As I suspected the problem is something else. It's this...

TypeScript is picking the second overload, but why? Let's remove the second overload and let's keep only the model overload then this happens...

![image](https://user-images.githubusercontent.com/30295578/126847452-ca97d77b-9aa9-42ac-b4a6-871469198cf9.png)

What typescript is saying is `typeof userModel` does not extend `Model<any, any, any, any>`, we can even check it...

![image](https://user-images.githubusercontent.com/30295578/126847568-ef97462e-b5c6-49f5-bfd7-2d7b4da647e6.png)

Which basically means `userModel` can't be assigned to `Model<any, any, any, any>`...

![image](https://user-images.githubusercontent.com/30295578/126847657-dae7f9c3-0f68-4c9f-bf38-6fc07780cb71.png)

And why is that? Because...

![image](https://user-images.githubusercontent.com/30295578/126847750-b1d265d8-771d-44b0-a3ef-e11c2a6ba5c4.png)

So perhaps what you were trying to do is make the `never` to be `EventObject` so that it works...

![image](https://user-images.githubusercontent.com/30295578/126847793-2ac3cb25-b822-49a4-be82-fd2ef6216236.png)

But that could be the incorrect way of solving it. Here are two possible reasons of the original problem -
1. `createModel` definition is bad, it should never return `Model<X, never>`
2. `Model` type definition is bad, `Model<X, never>` should extend `Model<any, any>` (or `UnknownModel` to be precise)

Now if the reason is former than your fix is correct to never return `Model<X, never>` but I'm suspicious the problem is latter. Why? I had a very similar problem with `Behavior`...

`createBehaviorFromPromise` returns `Behavior<never, T>` but it wasn't assignable to `UnknownBehavior`. Here too there could be the same two reasons - either `createBehaviorFromPromise` is bad or `Behavior` is bad. It was very apparent from the definition of `Behavior` that a promise behavior should be `Behavior<never, T>` - so `createBehaviorFromPromise` was right and the problem was with `Behavior`. So I made relax all the call signatures in `Behavior` so that the it's less strict and [assignability works](https://github.com/devanshj/txstate/commit/f92b56be918e2fd62db4cdbc11f2ec2692511164#diff-cd8075578a54f2b8997de6a2c06bccf76356e9c93bb5c153cefbd91fb33a1cefR17) (the changes are in the same commit you can take a quicklook)

Now in our case even if let's say the reason is latter, xstate is way too filled with "type-debt" to even fix it (I did attempt to change the variance of things all the way down but no luck). So I'd say let's just pretend the problem is former and never produce `Model<X, never>`.

But your definition of cast will make cast no longer cast (I'll explain later why `never extends X` holds), cast isn't the problem here and changing it will result in bigger problems so that's an absolute no-go. What you should do is this....

```ts
IsNever<TComputedEvent> extends true ? EventObject : Cast<TComputedEvent, EventObject>
```

I hope it helps! And sorry/welcome for the long explanation :P

Reply:
Okay so why `never extends X` holds? (X can be anything it'd always hold).

A type is basically a set of values assignable to it. So the set equivalent of `"a" | "b" | "c"` is `{ "a", "b", "c" }`. So when we want to check if A extends B what we're asking is - is a value of type B assignable to a value of type A (this is by the definition of `extends`). Which would be asking is the set equivalent of type B a subset of set equivalent of type A.

So when you look at the following code...
```ts
let a: "a" = "a";
let aOrBOrC: "a" | "b" | "c" = a;
```
it makes sense because `{ "a" }` is a subset of `{ "a", "b", "c" }`. Now `Ø` (null set) is also a subset of `{ "a", "b", "c" }` so the type equivalent of `Ø` should also be assignable `"a" | "b" | "c"`. Now the type equivalent of `Ø` in TypeScript is `never` which is called the "bottom type". Hence just like how `Ø ∪ X` is `X` similarly `never | X` is `X`.

Okay now fuck all this xD let me aslo explain you intuitively.

```ts
let x: "a" | "b" | "c" = {} as "a" | "b" | "c"   // as it is
let x: "a" | "b" | "c" = {} as "a" | "b"         // pop
let x: "a" | "b" | "c" = {} as "a"               // pop pop
let x: "a" | "b" | "c" = {} as never | "a"       // same as above
let x: "a" | "b" | "c" = {} as never             // pop pop pop
````

This is also the reason the lower bound in `type Extract<T, U> = T extends U ? T : never` is `never` so that `Extract<"a", "a">` gives `never`.

You can even think of it in terms of control flow analysis.
```
let x: "a" | "b" | "c" = {} as "a" | "b" | "c"
if (x === "c") return;
x // "a" | "b"
if (x === "b") return;
x // "a"
if (x === "a") return;
x // never;

let event: { type: string };
event = x; // now it seems should give an compile error because of the following line
event.type.toUpperCase();
// but this would never execute as the control flow never reaches here
// so it's okay to assign `never` to `{ type: string }`
```

What's why when you return `never` from a function typescript understands that the control flow will [never go past that](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3iAHgSwwC54AKASngF4A+eFEANxBgG4AoUSWBZdLKYuSq0A3gF82bXAXLso5IA). So essentially a function that returns `never` never actually returns. And hence again it's safe to write `let x: { type: string } = exit()` because exit would never return so who cares about `x`

Again sorry/welcome for the long explanation xD

Reply:
I would very much like to look at this before we merge this (even if only to level up my TS)... but it doesn't seem to be a lightweight material for a weekend 😬 I promise to take a look at this on Monday 🙏

Reply:
@devanshj I'm pretty sure we want the events for `Model` to never be `never`, because a model without events specified should result in a machine that can accept an `EventObject`. 

This PR implements your suggestion: https://github.com/statelyai/xstate/pull/2266/commits/458abd5750b0f6858c48d3fe779f987e2c64e2a0

Reply:
> because a model without events specified should result in a machine that can accept an EventObject

Why would you say that? A machine based on `Model<{}, { type: "FOO" } | { type: "BAR" }>` only takes events that are assignable to `{ type: "FOO" } | { type: "BAR" }`. Then in the same way a machine based on `Model<{}, never>` should take event assignable to `never` which basically means it should take no events.
Ah wait gotcha, you mean that if `createModel` has no event creators then we should fallback to `{ type: string }`, well if that's a conscious choice then it's okay.

But I still think `Model<{ foo: string }, never>` should extend `Model<unknown, { type: string }>` the root problem is that [`Action<{ foo: string }, never>` is not assignable to `UnknownAction` which is `Action<unknown, UnknownEvent>`](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggxsAlgewHYB4DCAaKBRKCAD2AlQBMBnKAVVQGtVkB3VPANzOAD4oBeKAAo4aUiQBcUHIU6pgkvAEp+vdskTkA3AChQkWgyasOXflADeUPREmVgAJ0SoA5lAC+u8NDqMWqeEhoZgEoGACuhn64PkZssjza2gA2EMBQRJIxfiFBAuZuUACG1DkYlgBmyMi2Dk6ubrioEJz23EA) which makes me think [`UnknownAction` should be `Action<never, never>` instead](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggxsAlgewHYB4DCAaKBRKCAD2AlQBMBnKAVVQGtVkB3VPANzOAD4oBeKAAo4aUiQBcUHIU6pgkvAEp+vdskTkA3AChQkWgyasOXflADeUPREmVgAJ0SoA5lAC+u8NDqMWqeEhoZgEoGKgQnPa44ZHc2toANhDAUESSPkb+CKFm5m5QAIbUIWjolgBmyMi2Dk6ubtEREPbcQA). So maybe instead of writing `Model<any, any>` one should write `Model<any, never>` idk. I'm not sure what's the "principally correct" way to solve it I haven't thought deeply about this.

But all this aside I given you're consciously making that choice then it's okay to do what you're doing with `CastIfNever`. Though to nitpick I'd suggest you to inline the `IsNever` check and remove the `CastIfNever`. Having that type would encourage it's usage where it not might be apt like a future person would go "Hmm Cast is not working, I see there is a CastIfNever" \*slaps that\* "Ha now it's working cool" and calls it a day xD instead of looking into the problem. But feel free to take your call, if you'd like to keep `CastIfNever` do that.

Reply:
Ok, I think I have finally digested this thread. 

Would it be correct to say that David's problem was caused by the variance issues? `Model["assign"]` is a function so its parameter types are contravariant. This should not result in an error under `strictFunctionTypes: false` which we currently have implicitly configured - we plan to change this and enable strict mode in the whole repo but this has not been done yet. I don't get why both you and David have seen this error and why I can't see it on my machine, nor CI fails. That being said - it's a good thing that this has been caught now but I don't understand why it was caught 😬 

I also quite don't understand the `UnknownBehavior` fix - I've tried to replicate that [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgdgazgewO5wKIDcJ2FAXigG8pRIAuKAZ2ACcBLOAcygF8AoD86AWWQAmEADYAeACoBhZHggAPYABoo47LnzzguAdViIU6NXgB8JDlAtQAhtWoNmcKgAoAxsDlUpMrQoCUhUyxkBgFzS2ptZwgsTyNgfwJA4NDObnBoeCQ0OH4hYUIoXJFRAFd9bOVMg0wcEy4XGVooAFtBESoq7KL8omI2a11u0VIGsq06KjgS5oAjCDp2ZTho+eMuDiA) and it doesn't type check (which makes sense to me with `strictFunctionTypes: true`). How your Behavior/UnknownBehavior is different from this?

---

In general, I agree with @devanshj - this statement seems like an off choice:
> model without events specified should result in a machine that can accept an EventObject.

If the whole point of types is to make our apps safer then we shouldn't produce unsafe variants by default. If the consumer doesn't want to type something, they are in a rush or having trouble doing so, then they should **explicitly** annotate stuff as being unsafe (with `any`). This IMHO should always be a conscious decision made by a user and not something that we promote by effectively making that a default that can be produced unintentionally.

Reply:
> If the whole point of types is to make our apps safer then we shouldn't produce unsafe variants by default. If the consumer doesn't want to type something, they are in a rush or having trouble doing so, then they should **explicitly** annotate stuff as being unsafe (with `any`). This IMHO should always be a conscious decision made by a user and not something that we promote by effectively making that a default that can be produced unintentionally.

Okay, that makes sense. I'll make the appropriate changes.

Reply:
> Would it be correct to say that David's problem was caused by the variance issues? Model["assign"] is a function so its parameter types are contravariant.

Right. Idr if it was the `assign` that caused the issue or not but you're probably right.

> I don't understand why it was caught 😬

Maybe our vscode was buggy and didn't pick the tsconfig? xD

>  I've tried to replicate that here

Not sure by "replicate" you meant replicate the bug or replicate the fix. If you meant the latter then nope xD Look closely the call signatures are different from `Behavior`, [here's how you fix it](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgdgazgewO5wKIDcJ2FAXigG8pRIAuKAZ2ACcBLOAcygF8AoDp4COgMwCGAY2gBZZABMIAGwA8AFQDCyPBAAewADRQF2XPg284k6rEQp0+vAD4SHKI6iDq1BszgAKYcHVVlqryaAJRUWMgMkg5O1LiSnhBY-tbAoVDhkRycHOTQ8EhocBLSMoRQxbJyAK4WhTr5lpg4tlzCqrRQALZSslQNhRWlRMRszmaDcqRtNbx0VHBVnQBGfOw6cIl8NlwcQA) (`type` also works but using `interface` just to be on a safer side). I tried applying this fix that day, either I applied it incorrectly or your repro is too minimal.